### DISCUSSION TITLE
Nested JSON field not updating in SQLAlchemy

### DISCUSSION THREAD

**Discussion by SanketDG** [sqlalchemy/sqlalchemy#6473] [Usage Questions] [answered]
Similar: https://github.com/sqlalchemy/sqlalchemy/issues/5218

I have a model attribute:
```
    status = Column(JSON(), nullable=True)
```

This is a nested JSON that is max 2 levels deep

How can I update a nested field in the JSON and commit changes?

I have tried all iterations of `dict.copy()`, `deepcopy`, using a brand new object, and nothing seems to work.

I do not want to make changes to `models.py`.
I am waiting to see if I have options/hacks right in the face, otherwise I will either go with:

- `sqlalchemy-json` library that can also deal with nested json
- `sqlachemy.ext.mutable`

**Comment by zzzeek** [comment, marked as answer]
hi -

solution 1: you would need to make use of [MutableDict](https://docs.sqlalchemy.org/en/14/orm/extensions/mutable.html#establishing-mutability-on-scalar-column-values) rather than using JSON() directly.    since you want to support mutiple levels, you would also coerce an incoming sub-dict to mutabledict, like:

```python
class MutableDict(Mutable, dict):
    @classmethod
    def coerce(cls, key, value):
        "Convert plain dictionaries to MutableDict."

        if not isinstance(value, MutableDict):
            if isinstance(value, dict):
                return MutableDict({
                    key: MuableDict(value) if isinstance(value, dict) else value for key, value in value.items()
                })

            # value is not a dict, this will raise
            return Mutable.coerce(key, value)
        else:
            return value

    def __setitem__(self, key, value):
        "Detect dictionary set events and emit change events."

        if isinstance(value, dict):
            value = MutableDict(value)
        dict.__setitem__(self, key, value)
        self.changed()
```

solution 2:   assign a copy of the dict *before* you change it:

```python
myobject.my_json_field = dict(my_object.my_json_field)
myobject.my_json_field['foo'] = 'bar'
```


solution 3: use flag_modified:

```python
from sqlalchemy.orm import attributes
myobject.my_json_field['foo'] = 'bar'
attributes.flag_modified(myobject, "my_json_field")

**Reply by SanketDG** [reply]
@zzzeek Thank you so much, I was foolishly deep copying the dictionary after making changes, not realizing how SQLAlchemy picks up changes. All of these are great suggestions, I have for now gone with _2_

Also thank you for all of the work you have been doing.
